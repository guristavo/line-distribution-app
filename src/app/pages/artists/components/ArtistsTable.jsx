import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

// Import shared components
import { FavoriteIcon, LoadingIcon, Icon } from '../../../common';
// Import utils
import enums from '../../../../utils/readable-enums';

const NUMBER_OF_COLUMNS = '7';

const ArtistsTable = ({
  artists,
  favoriteAction,
  pending,
  rowAction,
  searchQuery,
  showFavoriteArtistsOnly,
  user,
}) => {
  // Filter artists
  let filteredArtists = artists;
  if ((searchQuery && filteredArtists[0]) || showFavoriteArtistsOnly) {
    filteredArtists = _.filter(artists, o => {
      let include = false;
      // If displaying only favorite artists
      if (showFavoriteArtistsOnly) {
        include = user.favoriteArtists[o.id] !== undefined;
      }
      // If displaying only matching queries
      if (searchQuery && filteredArtists[0]) {
        include = o.query.includes(searchQuery);
      }

      return include;
    });
  }

  // Message to be display when table has no rows
  const emptyTableMessage =
    searchQuery.length > 0
      ? 'No artists available within your search'
      : 'No artists available';

  // Row Fallback when loading or no results
  const rowFallback = () => {
    if (pending) {
      return (
        <tr>
          <td colSpan={NUMBER_OF_COLUMNS}>
            <LoadingIcon />
          </td>
        </tr>
      );
    }
    return (
      <tr>
        <td colSpan={NUMBER_OF_COLUMNS}>{emptyTableMessage}</td>
      </tr>
    );
  };

  const hasActiveFilters = searchQuery.length > 0 || showFavoriteArtistsOnly;

  return (
    <Fragment>
      {pending ? (
        <p className="italic">Loading...</p>
      ) : (
        <p className="italic">
          Displaying {filteredArtists.length} artists
          {hasActiveFilters ? ' within search' : ''}
          ...
        </p>
      )}

      <table className="table">
        <thead>
          <tr>
            <th className="artists-table__favorite-col">Favorite</th>
            <th>Name</th>
            <th>Agency</th>
            <th>Genre</th>
            <th>Units</th>
            <th>Members</th>
            <th>Disbanded</th>
          </tr>
        </thead>
        <tbody onClick={rowAction}>
          {filteredArtists.length > 0
            ? filteredArtists.map(entry => {
                const key = `all-artists-${entry.id}`;
                const id = `${entry.id}:::${entry.unitIds[0]}`;
                return (
                  <tr key={key} id={id}>
                    <td
                      className="artists-cell-favorite"
                      onClick={() => favoriteAction(entry.id)}
                    >
                      <FavoriteIcon
                        action={() => {}}
                        id={entry.id}
                        size="12"
                        state={
                          user.favoriteArtists && user.favoriteArtists[entry.id]
                        }
                      />
                    </td>
                    <td className="no-break">
                      {entry.name}{' '}
                      {entry.private ? (
                        <Icon
                          type="private"
                          color="red"
                          title="Private"
                          inline
                        />
                      ) : null}
                    </td>
                    <td className="no-break">{entry.agency}</td>
                    <td>{enums.GENRES[entry.genre]}</td>
                    <td>{entry.unitIds.length}</td>
                    <td>
                      {entry.members.map(member => member.name).join(', ')} (
                      {entry.members.length})
                    </td>
                    <td className="center">
                      {entry.disbanded ? (
                        <Icon
                          type="grave"
                          color="purple"
                          title="Disbanded"
                          inline
                        />
                      ) : null}
                    </td>
                  </tr>
                );
              })
            : rowFallback()}
        </tbody>
      </table>
    </Fragment>
  );
};

ArtistsTable.propTypes = {
  artists: PropTypes.array.isRequired,
  favoriteAction: PropTypes.func.isRequired,
  pending: PropTypes.bool,
  rowAction: PropTypes.func.isRequired,
  searchQuery: PropTypes.string,
  showFavoriteArtistsOnly: PropTypes.bool,
  user: PropTypes.object.isRequired,
};

ArtistsTable.defaultProps = {
  pending: false,
  searchQuery: '',
  showFavoriteArtistsOnly: false,
};

export default ArtistsTable;
