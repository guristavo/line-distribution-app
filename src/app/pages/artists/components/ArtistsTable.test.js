import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ArtistsTable from './ArtistsTable';

const mockData = {
  artists: [
    {
      id: '-LZJ7AQQhyOk5w_TAD4u',
      agency: 'CUBE',
      createdBy: 'hbFlRswbZkepQfaONzoyB6EuJSA2',
      disbanded: false,
      genre: 'K-POP',
      members: [
        {
          id: '-LZJ7AH22seM59PQ977M',
          name: 'Miyeon',
          color: '23',
        },
        {
          id: '-LZJ7AH7krbGtR-yZyhT',
          name: 'Minnie',
          color: '2',
        },
        {
          id: '-LZJ7AHDQIEw5VM1mI3C',
          name: 'Soojin',
          color: '6',
        },
        {
          id: '-LZJ7AGyuXFDjXz1ECb2',
          name: 'Soyeon',
          color: '18',
        },
        {
          id: '-LZJ7AHM3LgaoxdPJphc',
          name: 'Yuqi',
          color: '14',
        },
        {
          id: '-LZJ7AHNENioOO_v3cE5',
          name: 'Shuhua',
          color: '8',
        },
      ],
      name: '(G)I-DLE',
      modifiedBy: null,
      otherNames: 'gidle g-i-dle g-idle g i dle',
      private: false,
      query:
        '(g)i-dle gidle g-i-dle g-idle g i dle miyeon minnie soojin soyeon yuqi shuhua cube',
      unitIds: ['-LZJ7AZJOR0ehcrPecgO'],
    },
  ],
  userFavoriteArtists: {
    abc123: true,
    def456: true,
  },
};

describe('pages/Artists/ArtistsTable', () => {
  let SPY;
  const NOOP = () => {};

  beforeEach(() => {
    SPY = jest.fn();
  });

  it('renders', () => {
    const { asFragment } = render(
      <ArtistsTable
        artists={mockData.artists}
        favoriteAction={NOOP}
        rowAction={NOOP}
        userFavoriteArtists={mockData.userFavoriteArtists}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
