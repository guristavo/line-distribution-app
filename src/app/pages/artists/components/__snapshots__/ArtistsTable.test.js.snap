// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Artists/ArtistsTable renders 1`] = `
<DocumentFragment>
  <p
    class="italic"
  >
    Displaying 0 artists...
  </p>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th
          class="artists-table__favorite-col"
        >
          Favorite
        </th>
        <th>
          Name
        </th>
        <th>
          Agency
        </th>
        <th>
          Genre
        </th>
        <th>
          Units
        </th>
        <th>
          Members
        </th>
        <th>
          Disbanded
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colspan="7"
        >
          No artists available
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
